docs/writing-tests/test-runner.md

```diff
@@ -0,0 +1,293 @@
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+<div class="aside">
+
+💡 Except for the `setup` function, all other functions run asynchronously. Both `preRender` and `postRender` functions include two additional arguments, a [Playwright page](https://playwright.dev/docs/pages) and a context object which contains the `id`, `title`, and the `name` of the story.
+
+</div>
+
+When the test runner executes, your existing tests will go through the following lifecycle:
+
+- The `setup` function is executed before all the tests run.
+- The context object is generated containing the required information.
+- Playwright navigates to the story's page.
+- The `preRender` function is executed.
+- The story is rendered, and any existing `play` functions are executed.
+- The `postRender` function is executed.
+
+### Stories.json mode
+
+When testing a local Storybook, the test runner transforms your story files into tests. For a remote Storybook, it uses the Storybook's [stories.json](../configure/overview.md#feature-flags) file (a static index of all the stories) to run the tests. Suppose you run into a situation where the local and remote Storybooks appear out of sync, or you might not even have access to the code. In that case, the `stories.json` file is guaranteed to be the most accurate representation of the deployed Storybook you are testing. To test a local Storybook using this feature, use the `--stories-json` flag as follows:
+
+<!-- prettier-ignore-start -->
+
+<CodeSnippets
+  paths={[
+    'common/storybook-test-runner-with-stories-json.yarn.js.mdx',
+    'common/storybook-test-runner-with-stories-json.npm.js.mdx',
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+<div class="aside">
+
+💡 The `stories.json` mode is not compatible with watch mode.
+
+</div>
+
+To disable it, use the `--no-stories-json` flag:
+
+<!-- prettier-ignore-start -->
+
+<CodeSnippets
+  paths={[
+    'common/storybook-test-runner-disable-stories-json.yarn.js.mdx',
+    'common/storybook-test-runner-disable-stories-json.npm.js.mdx',
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+---
+
+## Troubleshooting
+
+### The test runner seems flaky and keeps timing out
+
+If your tests time out with the following message:
+
+<!-- prettier-ignore-start -->
+
+<CodeSnippets
+  paths={[
+    'common/storybook-test-runner-error.js.mdx',
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+It might be that Playwright couldn't handle testing the number of stories you have in your project. Perhaps you have a large number of stories, or your CI environment has a really low RAM configuration. In such cases, you should limit the number of workers that run in parallel by adjusting your command as follows:
+
+```json
+{
+  "scripts": {
+    "test-storybook:ci": "concurrently -k -s first -n \"SB,TEST\" -c \"magenta,blue\" \"yarn build-storybook --quiet && npx http-server storybook-static --port 6006 --silent\" \"wait-on tcp:6006 && yarn test-storybook --maxWorkers=2\""
```

Comment:
The `--maxWorkers=2` flag is the only relevant bit here, right? I think we should avoid showing this complex recipe, if we can help it.

Reply:
Snippet was updated accordingly. Thanks for this